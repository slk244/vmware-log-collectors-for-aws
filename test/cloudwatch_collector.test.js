/*
Copyright 2018 VMware, Inc.
SPDX-License-Identifier: MIT
*/

const { lintTestEnv, sendLogsAndVerify } = require('../test/helper.test');
const { createSample1 } = require('../test/cloudwatch_testdata');
const { kubernetesLogSample } = require('../test/cloudwatch_testdata');

const {
  CloudWatchHttpCollector,
  CloudWatchKafkaCollector,
  processLogTextAsJson,
} = require('../src/index');

describe('processLogTextAsJson', () => {
  it('should process log text in JSON format', () => {
    const textJson = processLogTextAsJson(`
      { 
        "field1": { 
          "field11": "value11",
          "field12": "value12"
        },
        "field2": "value2", 
        "field3": 3,
        "field4": {
          "field41": ["a", "b", "c"]
        }
      }
    `);
    expect(textJson).toMatchSnapshot();
  });

  it('should process log text not in JSON format', () => {
    const textJson = processLogTextAsJson(' { "field2": "value2", field3: "value3" } ');
    expect(textJson).toEqual({});
  });
});

describe('CloudWatchKafkaCollector', () => {
  const collector = new CloudWatchKafkaCollector(lintTestEnv, new Map());

  it('processLogsJson', () => {
    const logsJson = createSample1();
    collector.processLogsJson(logsJson);
    expect(logsJson.subscriptionFilters).toBe(undefined);
    expect(logsJson.logEvents).toBe(undefined);
    expect(logsJson).toMatchSnapshot();
  });
});

describe('CloudWatchHttpCollector', () => {
  const tagRegexMap = new Map(Object.entries({
    tag1: new RegExp('c(k+)c', 'i'),
  }));
  const collector = new CloudWatchHttpCollector(lintTestEnv, tagRegexMap);

  describe('processLogsJson', () => {
    it('should process logs JSON', () => {
      const logsJson = createSample1();
      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });

    it('should merge in JSON fields if log text is in JSON format', () => {
      const logsJson = {
        logEvents: [
          {
            id: 'id1',
            text: ' { "timestamp": "1234", "field3": "value3" } ',
          },
        ],
      };

      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });

    it('should merge in JSON fields and extract tags', () => {
      const logsJson = {
        logEvents: [
          {
            id: 'id1',
            text: ' { "field2": "value2", "field3": "abCkKKkcde" } ',
          },
        ],
      };

      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });

    it('should extract tags from text string', () => {
      const logsJson = {
        logEvents: [
          {
            id: 'id1',
            text: 'abCkKKkcde',
          },
        ],
      };

      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });

    it('should extract tags from text array', () => {
      const logsJson = {
        logEvents: [
          {
            id: 'id1',
            text: [{ field1: 'abckkkkcde' }],
          },
        ],
      };

      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });

    it('should process kubernetes logs', () => {
      const logsJson = kubernetesLogSample();
      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });

    it('should modify unsupported special characters in keys to support LINT', () => {
      const logsJson = {
        logEvents: [
          {
            id: 'id1',
            text: ' { "field2/": "value2", "field3.": "abCkKKkcde" } ',
          },
        ],
      };

      collector.processLogsJson(logsJson);
      expect(logsJson).toMatchSnapshot();
    });
  });

  describe('sendLogs', () => {
    it('should send request to the HTTP stream', (done) => {
      const logsJson = {
        logEvents: [
          {
            id: 'id1',
            field1: 'value1',
          },
        ],
      };

      const expectedReqHeaders = {
        reqheaders: {
          authorization: 'Bearer mocktoken',
          structure: 'cloudwatch',
          'content-type': 'application/json',
        },
      };

      const expectedReqBody = {
        logEvents: [
          {
            id: 'id1',
            field1: 'value1',
            log_type: 'aws_cloud_watch',
          },
        ],
        ingest_timestamp: 1538769915817,
      };

      sendLogsAndVerify(done, collector, logsJson, expectedReqBody, expectedReqHeaders);
    });
  });
});
