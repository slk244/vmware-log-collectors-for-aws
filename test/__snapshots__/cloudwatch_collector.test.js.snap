// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudWatchHttpCollector processLogsJson should extract tags from text array 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "id": "id1",
      "log_type": "aws_cloud_watch",
      "tag1": "ckkkkc",
      "text": Array [
        Object {
          "field1": "abckkkkcde",
        },
      ],
    },
  ],
}
`;

exports[`CloudWatchHttpCollector processLogsJson should extract tags from text string 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "id": "id1",
      "log_type": "aws_cloud_watch",
      "tag1": "CkKKkc",
      "text": "abCkKKkcde",
    },
  ],
}
`;

exports[`CloudWatchHttpCollector processLogsJson should merge in JSON fields and extract tags 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "field2": "value2",
      "field3": "abCkKKkcde",
      "id": "id1",
      "log_type": "aws_cloud_watch",
      "tag1": "CkKKkc",
      "text": " { \\"field2\\": \\"value2\\", \\"field3\\": \\"abCkKKkcde\\" } ",
    },
  ],
}
`;

exports[`CloudWatchHttpCollector processLogsJson should merge in JSON fields if log text is in JSON format 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "field3": "value3",
      "id": "id1",
      "log_type": "aws_cloud_watch",
      "text": " { \\"timestamp\\": \\"1234\\", \\"field3\\": \\"value3\\" } ",
      "timestamp": 1234,
    },
  ],
}
`;

exports[`CloudWatchHttpCollector processLogsJson should modify unsupported special characters in keys to support LINT 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "field2_": "value2",
      "field3_": "abCkKKkcde",
      "id": "id1",
      "log_type": "aws_cloud_watch",
      "tag1": "CkKKkc",
      "text": " { \\"field2/\\": \\"value2\\", \\"field3.\\": \\"abCkKKkcde\\" } ",
    },
  ],
}
`;

exports[`CloudWatchHttpCollector processLogsJson should process kubernetes logs 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "docker_container_id": "5b15e87886a7ca5f7ebc73a15aa9091c9c0f880ee2974515749e16710367462c",
      "id": "34051922485820161636472577410016310394916410432944603138",
      "kubernetes_container_name": "weave",
      "kubernetes_host": "kube-234",
      "kubernetes_labels_controller-revision-hash": "2720543195",
      "kubernetes_labels_name": "weave-net",
      "kubernetes_labels_pod-template-generation": "1",
      "kubernetes_master_url": "https://10.96.0.1:443/api",
      "kubernetes_namespace_name": "kube-system",
      "kubernetes_pod_id": "ac4bdfc1-9dc0-11e7-8b62-005056b549b6",
      "kubernetes_pod_name": "weave-net-4n4kc",
      "log_type": "aws_cloud_watch",
      "stream": "stderr",
      "text": "INFO: 2017/10/02 06:44:13.214543 Discovered remote MAC 62:a1:3d:f6:eb:65 at 62:a1:3d:f6:eb:65(kube-235)
",
    },
  ],
  "logGroup": "/aws/lambda/runbook",
  "logStream": "unittest8",
  "messageType": "DATA_MESSAGE",
  "owner": "792677813823",
}
`;

exports[`CloudWatchHttpCollector processLogsJson should process logs JSON 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logEvents": Array [
    Object {
      "id": "34051922485820161636472577410016310394916410432944603138",
      "log_type": "aws_cloud_watch",
      "text": "2018-05-21T22:16:53.934Z	aab15e49-5d44-11e8-a7db-afa704cec35a	Status: 200
",
    },
    Object {
      "id": "001",
      "log_type": "aws_cloud_watch",
      "message": "message of 001",
      "text": "text of 001",
    },
    Object {
      "id": "34051922485820161636472577410016310394916410432944603138",
      "log_type": "aws_cloud_watch",
      "text": Array [
        Object {
          "fielda": "fielda",
        },
      ],
    },
  ],
  "logGroup": "/aws/lambda/runbook",
  "logStream": "unittest8",
  "messageType": "DATA_MESSAGE",
  "owner": "792677813823",
}
`;

exports[`CloudWatchKafkaCollector processLogsJson 1`] = `
Object {
  "ingest_timestamp": 1538769915817,
  "logGroup": "/aws/lambda/runbook",
  "logStream": "unittest8",
  "logs": Array [
    Object {
      "id": "34051922485820161636472577410016310394916410432944603138",
      "log_type": "aws_cloud_watch",
      "text": "2018-05-21T22:16:53.934Z	aab15e49-5d44-11e8-a7db-afa704cec35a	Status: 200
",
    },
    Object {
      "id": "001",
      "log_type": "aws_cloud_watch",
      "message": "message of 001",
      "text": "text of 001",
    },
    Object {
      "id": "34051922485820161636472577410016310394916410432944603138",
      "log_type": "aws_cloud_watch",
      "text": Array [
        Object {
          "fielda": "fielda",
        },
      ],
    },
  ],
  "messageType": "DATA_MESSAGE",
  "owner": "792677813823",
  "structure": "cloudwatch",
}
`;

exports[`processLogTextAsJson should process log text in JSON format 1`] = `
Object {
  "field1_field11": "value11",
  "field1_field12": "value12",
  "field2": "value2",
  "field3": 3,
  "field4_field41_0": "a",
  "field4_field41_1": "b",
  "field4_field41_2": "c",
}
`;
